packages/utils/src/lib/reports/generate-md-report.ts

```diff
@@ -1,171 +1,150 @@
-    `${FOOTER_PREFIX}${SPACE}${link(README_LINK, 'Code PushUp')}`,
-  );
+  return new MarkdownDocument()
+    .heading(HIERARCHY.level_1, reportHeadlineText)
+    .$if(report.categories.length > 0, doc =>
+      doc.$concat(
+        categoriesOverviewSection(report),
+        categoriesDetailsSection(report),
+      ),
+    )
+    .$concat(auditsSection(report), aboutSection(report))
+    .rule()
+    .paragraph(md`${FOOTER_PREFIX} ${md.link(README_LINK, 'Code PushUp')}`)
+    .toString();
 }
 
-export function auditDetailsIssues(issues: Issue[] = []) {
+export function auditDetailsIssues(
+  issues: Issue[] = [],
+): MarkdownDocument | null {
   if (issues.length === 0) {
-    return '';
+    return null;
   }
-  const detailsTableData = {
-    title: 'Issues',
-    columns: issuesTableHeadings,
-    rows: issues.map(
-      ({ severity: severityVal, message, source: sourceVal }: Issue) => {
-        const severity = `${severityMarker(severityVal)} <i>${severityVal}</i>`;
-
-        if (!sourceVal) {
-          return { severity, message, file: '', line: '' };
-        }
-        // TODO: implement file links, ticket #149
-        const file = `<code>${sourceVal.file}</code>`;
-        if (!sourceVal.position) {
-          return { severity, message, file, line: '' };
-        }
-        const { startLine, endLine } = sourceVal.position;
-        const line = `${startLine || ''}${
-          endLine && startLine !== endLine ? `-${endLine}` : ''
-        }`;
-        return { severity, message, file, line };
-      },
-    ),
-  };
-
-  return tableSection(detailsTableData);
+  return new MarkdownDocument().heading(HIERARCHY.level_4, 'Issues').table(
+    [
+      { heading: 'Severity', alignment: 'center' },
+      { heading: 'Message', alignment: 'left' },
+      { heading: 'Source file', alignment: 'left' },
+      { heading: 'Line(s)', alignment: 'center' },
+    ],
+    issues.map(({ severity: level, message, source }: Issue) => {
+      const severity = md`${severityMarker(level)} ${md.italic(level)}`;
+
+      if (!source) {
+        return [severity, message];
+      }
+      // TODO: implement file links, ticket #149
+      const file = md.code(source.file);
+      if (!source.position) {
+        return [severity, message, file];
+      }
+      const { startLine, endLine } = source.position;
+      const line = `${startLine || ''}${
+        endLine && startLine !== endLine ? `-${endLine}` : ''
+      }`;
+      return [severity, message, file, line];
+    }),
+  );
 }
 
-export function auditDetails(audit: AuditReport) {
+export function auditDetails(audit: AuditReport): MarkdownDocument {
   const { table, issues = [] } = audit.details ?? {};
   const detailsValue = auditDetailsAuditValue(audit);
 
   // undefined details OR empty details (undefined issues OR empty issues AND empty table)
-  if (issues.length === 0 && table == null) {
-    return section(detailsValue);
+  if (issues.length === 0 && !table?.rows.length) {
```

Comment:
⚙️ TypeScript: Wouldn't it be more consistent to use `length === 0`?
```suggestion
  if (issues.length === 0 && table?.rows.length === 0) {
```

Reply:
The trouble is the type of `table?.rows.length` isn't `number`, it's `number | undefined`. Missing tables are meant to be treated the same as empty tables, which works with falsy check (`!undefined` and `!0` are both true), but not when comparing against 0 (`0 === 0` is true, but `undefined === 0` is false).